airbyte-integrations/connectors/source-paypal-transaction/source_paypal_transaction/source.py

```diff
@@ -0,0 +1,394 @@
+        **kwargs,
+    ):
+
+        # Initiate data from config
+        if config:
+            start_date = isoparse(config["start_date"])
+            end_date_str = config.get("end_date")
+            end_date = isoparse(end_date_str) if end_date_str else None
+            is_sandbox = config["is_sandbox"]
+
+        self.is_sandbox = is_sandbox
+        self.start_date = start_date
+        self._end_date = end_date
+        self.is_sandbox = is_sandbox
+
+        self._validate_input_dates()
+
+        super().__init__(**kwargs)
+
+    @property
+    def end_date(self):
+        """Return initiated end_date or now()"""
+        now = datetime.now().replace(microsecond=0).astimezone()
+        if not self._end_date or self._end_date > now:
+            # If no end_date or end_date is in future then return now:
+            return now
+        else:
+            return self._end_date
+
+    def _validate_input_dates(self):
+
+        # Validate input dates
+        if self.start_date > self.end_date:
+            raise Exception(f"start_date {self.start_date.isoformat()} is greater than end_date {self.end_date.isoformat()}")
+
+        current_date = datetime.now().replace(microsecond=0).astimezone()
+        current_date_delta = current_date - self.start_date
+
+        # Check for minimal possible start_date
+        if current_date_delta > timedelta(**self.start_date_min):
+            raise Exception(
+                f"Start_date {self.start_date.isoformat()} is too old. "
+                f"Min date limit is {self.start_date_min} before now:{current_date.isoformat()}."
+            )
+
+        # Check for maximum possible start_date
+        if current_date_delta < timedelta(**self.start_date_max):
+            raise Exception(
+                f"Start_date {self.start_date.isoformat()} is too close to now. "
+                f"Max date limit is {self.start_date_max} before now:{current_date.isoformat()}."
+            )
+
+    @property
+    def url_base(self) -> str:
+
+        return f"{get_endpoint(self.is_sandbox)}/v1/reporting/"
+
+    def request_headers(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> Mapping[str, Any]:
+
+        return {"Content-Type": "application/json"}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+
+        json_response = response.json()
+        if self.data_field is not None:
+            data = json_response.get(self.data_field, [])
+        else:
+            data = [json_response]
+
+        for record in data:
+            # In order to support direct datetime string comparison (which is performed in incremental acceptance tests)
+            # convert any date format to python iso format string for date based cursors
+            self.update_field(record, self.cursor_field, lambda date: isoparse(date).isoformat())
+            yield record
```

Comment:
transactions and balances have slightly different structure. maybe it's better to implement `parse_response` for each stream separately ?

Reply:
maybe it's more related to `update_field` and `get_field` methods. Could you add some comments with explanation of `if not isinstance(field_path, List):` and why we need it

Reply:
field_path accept cursor attr which can be a string or a list of strings (for nested dicts), I updated type annotation for this argument